Implement minus for DateTime64

liyinsg opened the pull request:
Add minus operator support for DateTime64, resolves #63096

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add minus operator support for DateTime64, to allow subtraction between DateTime64 values, as well as DateTime

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory
and will run independently of the checks below:

- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64
- [ ] <!---ci_exclude_release--> Exclude: All with release
- [ ] <!---ci_exclude_debug--> Exclude: All with debug
---
- [ ] <!---ci_include_uzz--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, BuzzHouse, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

liyinsg committed:
Implement minus for DateTime64

Add minus operator support for DateTime64, resolves #63096

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #74482 from liyinsg/ClickHouse:feat-datatime64-subtraction -->
*This is an automated comment for commit 13ceb313576cea9e09b32d9c88ab8418f9b29978 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/74482/13ceb313576cea9e09b32d9c88ab8418f9b29978/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/0b03726596237f9de1e867c83dc6bb9f3f171ef3/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

liyinsg committed:
update document

tavplubix commented on tests/queries/0_stateless/02156_minus_op_with_datatime64.sql:
```diff
@@ -0,0 +1,53 @@
+-- DateTime64 (scale 0) - DateTime
+toDateTime64('2023-10-02 12:00:00', 0) - toDateTime('2023-10-01 11:00:00') AS result_no_materialize_3,
+
+-- DateTime64 (scale 3) - DateTime64 (scale 6)
+toDateTime64('2023-10-01 12:00:00', 3) - toDateTime64('2023-10-01 11:00:00', 6) AS result_no_materialize_4,
+
+-- DateTime64 (scale 6) - DateTime64 (scale 3)
+toDateTime64('2023-10-01 12:00:00', 6) - toDateTime64('2023-10-01 11:00:00', 3) AS result_no_materialize_5,
```
It would be nice to add a fraction part here, like `toDateTime64('2023-10-01 12:00:00.123456', 6) - toDateTime64('2023-10-01 11:00:00.123', 3)`

liyinsg replied:
Done

tavplubix commented on src/Functions/FunctionBinaryArithmetic.h:
```diff
@@ -1176,6 +1190,93 @@ class FunctionBinaryArithmetic : public IFunction
 
+    ColumnPtr executeDateTime64Subtraction(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_row_count) const
+    {
+        using OpImpl = DecimalBinaryOperation<Op, Decimal64, false>;
```
```suggestion
        using OpImpl = DecimalBinaryOperation<Op, Decimal64, /* check_overflow */ false>;
```

Why `false` btw?

liyinsg replied:
I checked the range of DateTime64, yes we need to check overflow. I've updated the code and test case to cover it, thanks

liyinsg committed:
address review concerns

1. Change code style according to the guideline
2. Check overflow for decimal operation

tavplubix commented on src/Functions/FunctionBinaryArithmetic.h:
```diff
@@ -1176,6 +1190,107 @@ class FunctionBinaryArithmetic : public IFunction
+    ColumnPtr executeDateTime64Subtraction(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type,
+                                           size_t input_row_count) const
+    {
+        using OpImpl = DecimalBinaryOperation<Op, Decimal64, /* check_overflow */ true>;
```
I think it's supposed to be controlled by a setting: https://github.com/ClickHouse/ClickHouse/blob/3f5006095f481f1716bccc4481ac7977ab7c5b0e/src/Functions/FunctionBinaryArithmetic.h#L1481
https://github.com/ClickHouse/ClickHouse/blob/6152f52ef04ed8553cb69d9cbe19d070009aa9e6/src/Core/Settings.cpp#L2941

liyinsg replied:
Thanks, addressed the concern by calling `helperInvokeEither` which checks `decimal_check_overflow`

liyinsg committed:
use check_decimal_overflow

liyinsg committed:
fix test failure caused by session_timezone

liyinsg committed:
refactor code with template lambda

tavplubix committed:
Merge branch 'master' into feat-datatime64-subtraction

tavplubix commented:
Stateless tests (release, ParallelReplicas, s3 storage) - https://github.com/ClickHouse/ClickHouse/issues/74547
Stateless tests (release, old analyzer, s3 storage, DatabaseReplicated) [2/2] - https://github.com/ClickHouse/ClickHouse/issues/74734
